### ISSUE TITLE
Core-AAM: AXPress and AXShowMenu on aria-haspopup is potentially misleading/confusing

### ISSUE THREAD

**Issue by joanmarie** [w3c/aria#570]
I came across this issue looking into why some of our CR tests were failing. The failing assertions are:

* property actions doesNotContain `AXShowMenu`
* property actions doesNotContain `AXPress`

These failing assertions are on a `button` with the value of `aria-haspopup` being `false` or invalid (empty string, bogus value like 'foo').

The Core AAM explicitly mentions `AXShowMenu` and `AXPress` in all the non-false mappings of `aria-haspopup`. And it does not mention these actions for the roles, e.g. [`button`](https://rawgit.com/w3c/aria/master/core-aam/core-aam.html#role-map-button). So it is not unreasonable for someone to conclude that "valid `aria-haspopup`" is what results in the need to implement the `AXShowMenu` and `AXPress` actions -- and then to take that assumption one step further, as was done in the assertions stated above, and conclude that "lack of valid `aria-haspopup`" means neither of these actions should be exposed.

However, in looking into the situation a bit more, my conclusion is that `AXShowMenu` and `AXPress` are completely unrelated to `aria-haspopup` and/or that additional guidance for implementors is needed.

1. According to Apple's developer docs, [`accessibilityPerformPress()`](https://developer.apple.com/reference/appkit/nsaccessibility/1526358-accessibilityperformpress) "Simulates clicking the element." One should "Use this method on elements that behave like buttons."

   I would argue that elements with role `button` behave like buttons. ;)

   Note that the Core AAM does also contain similar information, but it is down in the [Actions section](https://rawgit.com/w3c/aria/master/core-aam/core-aam.html#mapping_actions). The guidance there states, "If an AT triggers an AXPress action on an accessible object, the user agent should simulate a click on the DOM element which is mapped to that accessible object." If user agents follow that, then the make-the-popup-appear thing should JustWork™.

1. According to Apple's developer docs, [`accessibilityPerformShowMenu()`](https://developer.apple.com/reference/appkit/nsaccessibility/1532774-accessibilityperformshowmenu) "Displays this menu element." That sounds consistent with `aria-haspopup` being valid and non-`false`. But that same page states, "Use this method to display the contextual menu for the element." So that sounds like a right-click to me; not an action that would cause the popup associated with the element with `aria-haspopup` to become popped up.

1. According to a [comment](https://trac.webkit.org/browser/webkit/trunk/Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm?rev=215679#L1090) in the Mac port's WebKit accessibility code, "All elements should get ShowMenu and ScrollToVisible."

From all of the above, I'd be tempted to remove any mention of `AXShowMenu` and `AXPress` from `aria-haspopup` because they seem to belong on all interactive elements regardless of type.

In addition, it may be worth considering adding some text to the Core AAM regarding what user agents should do if an AT triggers an AXShowMenu action on an accessible object. Moreover, if my proposal to remove any mention of `AXShowMenu` and `AXPress` from `aria-haspopup` is rejected, I would argue that clarification text is essential.

**Comment by Joanmarie** [comment]
And the plot thickens....

The next batch of failing assertions are all associated with `combobox` elements with the value of `aria-haspopup` being valid and non-`false`. And the failing assertion is:

property actions contains `AXPress`

Given what I stated in my opening report, these failures struck me as implementation bugs. But then I noticed that in all the test cases, the pop up is popped up and the `combobox` is expanded. So I took a look at the code, in particular [`AccessibilityObject::supportsPressAction()`](https://trac.webkit.org/browser/webkit/trunk/Source/WebCore/accessibility/AccessibilityObject.cpp?rev=215679#L2291). As you'll see there, there is a comment which states:

```
// [Bug: 136247] Heuristic: element handlers that have more than one accessible descendant should not be exposed as supporting press.
```

and:

```
// Search within for immediate descendants that are static text. If we find more than one
// then this is an event delegator actionElement and we should expose the press action.
```

I've admittedly not debugged it yet, but the above comments and associated code seem to suggest that the test failures might not be a bug, but rather a feature. I'll ask Apple to find out which it is. But if it is indeed a feature, it contradicts what is specified in the Core AAM, which states

Expose `AXShowMenu` and `AXPress` actions

without any qualification. And even if it's a bug, there are still instances in which `AXPress` should not be exposed (according to the owners of the accessibility API being mapped as demonstrated through their user agent). Should those instances be described in the Core AAM?

**Comment by joanmarie** [comment]
Pull request #601 (committed) removed the AXPress action from the macOS mapping of aria-haspopup.

I still need to review everything to see what might remain before we can close this issue. That's on my todo-list.

**Comment by joanmarie** [comment]
> I still need to review everything to see what might remain before we can close this issue. That's on my todo-list.

Nothing is jumping out at me as still being broken. Closing as FIXED.
